### PR TITLE
fix: circular reference destructuring works with all models

### LINKED ISSUE #811: Regression in destructuring models from effects dispatch param
[Issuehunt badges](https://issuehunt.io/r/rematch/rematch/issues/811)

**Describe the bug**

Regression about destructuring dispatch models from function param.

**To Reproduce**
```js
export const count = {
	state: 0,
	reducers: {
		increment: s => s + 1,
	},
	effects: ({ count }) => ({
		async asyncIncrement() {
			await new Promise(resolve => {
				setTimeout(resolve, 1000)
			})

			count.increment()
		},
	}),
}
```

**Expected behavior**
Should work on the same way than not destructured like v1.

**Screenshots**
Screenshot 2020-08-27 at 22 34 53

**Additional context**
There's an issue related to circular reference that @tianzhich is looking for a solution, could be related.
Destructuring the models inside the effect function works good.
```js
export const count = {
	state: 0,
	reducers: {
		increment: s => s + 1,
	},
	effects: dispatch => ({
		async asyncIncrement() {
			const { count } = dispatch
			count.increment()
		},
	}),
}
```

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $25.80)

- [semoal semoal](https://issuehunt.io/u/semoal) ($25.80)

### Submitted pull Requests
- [#947 fix: circular reference destruction works with models](https://issuehunt.io/r/rematch/rematch/pull/947)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/rematch/rematch/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by sushantdhiman**
Fixes #811

Short version of this problem given below 

```js
/**---Problem---**/

const dispatch = {};

const gen = ({ a, b }) => {
  return {
    log: () => {
      console.log(a, b);
    }
  };
};

["a", "b"].forEach((item) => {
  const x = {};
  dispatch[item] = x;
  x['run' + item] = gen(dispatch);
});

dispatch["a"].runa.log();
dispatch["b"].runb.log();


/**--- Solution ---**/

const dispatch = {
  a: {},
  b: {}
};

const gen = ({ a, b }) => {
  return {
    log: () => {
      console.log(a, b);
    }
  };
};

["a", "b"].forEach((item) => {
  const x = dispatch[item] || {};
  dispatch[item] = x;
  x['run' + item] = gen(dispatch);
});

dispatch.a.runa.log();
dispatch.b.runb.log();
```

```sh
{ runa: { log: [Function: log] } } undefined
{ runa: { log: [Function: log] } } { runb: { log: [Function: log] } }

{ runa: { log: [Function: log] } } { runb: { log: [Function: log] } }
{ runa: { log: [Function: log] } } { runb: { log: [Function: log] } }
```

When effects method is called it can only access object references which were available to it at that time. I have separated model creation and effects creation process. Now correct models reference is available at the time of effects creation.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#811: Regression in destructuring models from effects dispatch param](https://issuehunt.io/repos/105411913/issues/811)
---
</details>

### CODE REVIEW

**Comment by semoal** [comment]
LGTM, let see what Zhi Tian says

File: packages/core/src/dispatcher.ts [hunk 1]
```diff
@@ -47,19 +47,17 @@ const createActionDispatcher = <
 
 /**
  * Creates a dispatcher object for a model - it contains a mapping from all
- * reducers and effects *names* to functions which dispatch their corresponding
- * actions.
+ * reducers to functions which dispatch their corresponding actions.
  */
-const createDispatcher = <
+export const createDispatcher = <
```
Comment (tianzhich) [inline packages/core/src/dispatcher.ts:52]: Maybe calling it `createReducerDispatcher` is better?

**Review by tianzhich — APPROVED** [review]
Except for one place about the function naming, all changes look great to me too.

**Comment by tianzhich** [comment]
Nice explanations here. Thanks for your contributions @sushantdhiman

File: packages/core/src/dispatcher.ts (hunk 1 above)
Comment (sushantdhiman) [reply packages/core/src/dispatcher.ts]: Done!

**Review by tianzhich — APPROVED** [review]

**Pull request merged**